### ISSUE TITLE
As a Firefox extension developer, I would like to install a temporary extension from a folder

### ISSUE THREAD

**Issue by seanpoulter** [SeleniumHQ/selenium#8357]
**Edit**: 2020-06-11 &ndash; ⚠️ This initial comment is not correct. I misunderstood the GeckoDriver API. I was looking at stale source code. 😓 With new knowledge and some guidance from the GeckoDriver team, I've proposed an approach to handle the "install addon" command from FirefoxDriver consistently for each language. See my [long comment](https://github.com/SeleniumHQ/selenium/issues/8357#issuecomment-643047209) below.

--

## 🐛 Bug Report

We can't install a temporary extension in Firefox using the JavaScript or DotNet bindings. The implementation is inconsistent across the languages. The GeckoDriver API expects a `path` as a String but we're sending an `addon`. See [mozilla/geckodriver - marionette.rs](https://github.com/mozilla/geckodriver/blame/17945ac838449dda1defc8cda0c6dadb042519b2/src/marionette.rs#L1319-L1328).

These two bindings need to be fixed:

|Language|Implementation|Authors / CC|
|-|-|-|
|JavaScript|[javascript/node/selenium-webdriver/firefox.js#L637-L657](https://github.com/SeleniumHQ/selenium/blob/master/javascript/node/selenium-webdriver/firefox.js#L637-L657)|@kvetko, @jleyba|
|DotNet|[dotnet/src/webdriver/Firefox/FirefoxDriver.cs#L213-L227](https://github.com/SeleniumHQ/selenium/blob/master/dotnet/src/webdriver/Firefox/FirefoxDriver.cs#L213-L227)|@jimevans|

These seem to be OK:

|Language|Implementation|
|-|-|
|Java|[java/client/src/org/openqa/selenium/firefox/FirefoxDriver.java#L235-L240](https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/firefox/FirefoxDriver.java#L235-L240)|
|Python|[py/selenium/webdriver/firefox/webdriver.py#L247-L264](https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/firefox/webdriver.py#L247-L264)|
|Ruby|[rb/lib/selenium/webdriver/firefox/bridge.rb#L34-L38](https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/firefox/bridge.rb#L34-L38)

## To Reproduce

Try to install an extension without building an .xpi or .zip file using dotnet and javascript bindings.

TODO - Script

## Expected behavior

I expect to be able to install a temporary extension in Firefox:
- From the JavaScript library using [`installAddon`](https://www.selenium.dev/selenium/docs/api/javascript/module/selenium-webdriver/firefox_exports_Driver.html#installAddon)
- From the DotNet library using [`InstallAddOn`](https://www.selenium.dev/selenium/docs/api/dotnet/?topic=html/T_OpenQA_Selenium_Firefox_FirefoxDriver.htm)
    - Note 1: This method is not in those docs?!
    - Note 2: `InstallAddOnFromFile` is also made redundant

## Test script or set of commands reproducing this issue

TODO - Include in example repo.

The steps to reproduce are:
* Create a minimal extension
    * **extension\manifest.json**
        ```json
        {
          "name": "Uh oh!",
          "description" : "Print a message to the console you may not see",
          "version": "0.1",
          "manifest_version": 2,
          "content_scripts": [
            {
              "matches": ["https://*.github.com/*"],
              "js": ["index.js"]
            }
          ]
        }
        ```
    * **extension\index.js**
        ```js
        console.log('We did it! ⭐️');
        ```
* Start Firefox and add an extension
    * **javascript/example.js**
        ```js
        const { Builder, By } = require('selenium-webdriver');
        const { Options } = require('selenium-webdriver/firefox');

        (async () => {

          const options =
            new Options()
              .setPreference('extensions.htmlaboutaddons.recommendations.enabled', false);

          const builder =
            new Builder()
            .forBrowser('firefox')
            .setFirefoxOptions(options);

          const driver = builder.build();

          const isTemporaryAddon = true;
          driver.installAddon('../extension/', isTemporaryAddon);

          await driver.get('https://github.com/SeleniumHQ/selenium');

          // Expected: A message to be printed on the console
          // Expected: The temporary extension appears in about:debugging > This Firefox > Temporary Extensions

        })();
        ```
* Install selenium-webdriver
    ```shell
    echo {} > package.json
    npm install --save-dev selenium-webdriver
    ```
* Run
    ```shell
    node javascript/example.js
    ```

## Environment

OS: macOS
Browser: Firefox
Browser version: 71.0
Browser Driver version: GeckoDriver for the last 2+ years
Language Bindings version: Latest
Selenium Grid version (if applicable): Not applicable

## Questions

1. Would you welcome a PR?
2. Do you want the issue or PR split by language?

**Comment by AutomatedTester** [comment]
Please can we have a 2 separate PRs for each language. I will try get them merged ASAP

**Comment by seanpoulter** [comment]
Will do. Thanks for the quick reply David.

**Comment by seanpoulter** [comment]
Hey @AutomatedTester, before I submit any code changes I wanted to check-in with a bit of a pivot. I've had a good chat with some of the Mozilla devs on [chat.mozilla.org](https://chat.mozilla.org) from the interop and Add-ons channels.

### Background

FirefoxDriver will dispatch a command to install an addon. GeckoDriver supports two variants of this command:
* { addon: string, temporary: Optional<boolean> }
* { path: string, temporary: Optional<boolean>

The "addon" is the Base64 encoded contents of a .xpi or .zip packed addon.
The "path" is a file path to a .xpi or .zip file. It can also be the `path` to an unpacked directory when:
- The `temporary` is set
- Firefox is configured to allow temporary extensions, either by changing the configuration or using the nightly or dev builds

### What did I get wrong in the issue description?

- GeckoDriver still supports both types of "install addon" command messages. I was looking at stale source code. See `AddonInstallParameters` in https://searchfox.org/mozilla-central/source/testing/geckodriver/src/command.rs.
- We don't want to move away from "addon" at all. @whimboo has explained that when we're handling .xpi or .zip files, it is best to use the "addon" message. This allows GeckoDriver and the browser to be on separate machines. Otherwise, we can use the "path" message.

### Proposed Changes

We'll want every language to use the same behaviour. When asked to install an addon, we should check the file type of the argument. When we have an .xpi or .zip file, we should send the command using the "addon"-type message. Otherwise, we'll use the "path".

To save space in the table below, I'll call it "🍍".

<table>
<thead><tr>
<th>Language</th>
<th>Current Behaviour</th>
<th>Proposed Changes</th>
</tr>
</thead>
<tbody>

<tr>
<td>.NET</td>
<td>See <a href="https://github.com/SeleniumHQ/selenium/blob/master/dotnet/src/webdriver/Firefox/FirefoxDriver.cs#L186-L227">FirefoxDriver.cs</a>
<ul>
<li><code>InstallAddOnFromFile(path)</code></li>
<li><code>InstallAddOn(base64encodedString)</code></li>
</ul></td>
<td><ul><li>Change the signature of <code>InstallAddOnFromFile</code> to include the optional <code>temporary</code> flag</li>
<li>Update <code>InstallAddOnFromFile</code>. 🍍</li>
<li><b>Breaking change</b>: Ideally we would rename <code>InstallAddOnFromFile</code> to <code>InstallAddOn</code></li>
</ul></td>
</tr>

<tr>
<td>Java</td>
<td>See <a href="https://github.com/SeleniumHQ/selenium/blob/master/java/client/src/org/openqa/selenium/firefox/FirefoxDriver.java#L235-L240">FirefoxDriver.java</a>
<ul>
<li><code>installExtension(Path path)</code></li>
</ul></td>
<td><ul>
<li>Deprecate <code>installExtension</code> or remove in favour of <code>installAddon</code>.</li>
<li>Add <code>installAddon(Path path)</code> and <code>installAddon(Path path, boolean temporary)</code> methods.</li>
<li>🍍</li>
</ul></td>
</tr>

<tr>
<td>JavaScript</td>
<td>See <a href="https://github.com/SeleniumHQ/selenium/blob/master/javascript/node/selenium-webdriver/firefox.js#L632-L652">firefox.js</a>
<ul>
<li><code>installAddon(path, temporary)</code></li>
</ul></td>
<td><ul>
<li>🍍</li>
</ul></td>
</tr>

<tr>
<td>Python</td>
<td>See <a href="https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/firefox/webdriver.py#L247-L264">firefox/webdriver.py</a>
<ul>
<li><code>install_addon(self, path, temporary=None)</code></li>
</ul></td>
<td><ul>
<li>🍍</li>
</ul></td>
</tr>

<tr>
<td>Ruby</td>
<td>See <a href="https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/firefox/bridge.rb#L34-L38">firefox/bridge.rb</a>
<ul>
<li><code>install_addon(path, temporary)</code></li>
</ul></td>
<td><ul>
<li>🍍</li>
</ul></td>
</tr>

</tbody>
</table>

We can also update the API docs to include a note that Firefox needs to be configured to allow `temporary` extensions.

**Comment by nickgaya** [comment]
I'd like to suggest a different approach.  Instead of checking the file extension or type, the API could allow the caller to specify whether the given path is local or remote.  If local, then the code should read the file and send it to the server using the `addon` parameter. If remote, the code should send the path to the server using the `path` parameter.

I think this would be more flexible and give the caller more control over the behavior.

**Comment by hheexx** [comment]
Without temporary flag we can't  add unsigned extensions

**Comment by bonigarcia** [comment]
Is there any update about this issue? Thanks.

**Comment by titusfortner** [comment]
Java fix is here - 932cbf96
I have a PR for .NET #10093 

I'm working to get everything implemented with `addon` instead of `path` for Selenium 4.2 so we don't have to duplicate Local File Detector implementations for using with Remote drivers. There's not much downside to just encoding it even when on a local machine, even if geckodriver doesn't need it.  JavaScript & .NET were already doing this, but [Java support](44f1fd983c290a5ae35181b9078157345cf1da66) and [Ruby support](b511a7dff84c4e8438f84d49ff9a2fda4ea8a306) are in trunk, and I have a PR for Python: #10092

**Comment by seanpoulter** [comment]
That makes sense with `addon` and both PRs look great. Thank you Titus!

**Comment by bonigarcia** [comment]
As the issue title says, I aim to install a temporary extension from a folder using Firefox. Using the Java bindings, this was already possible in Chrome using `--load-extension` as follows:

```
Path extension = Paths
        .get(ClassLoader.getSystemResource("web-extension").toURI());
ChromeOptions options = new ChromeOptions();
options.addArguments(
        "--load-extension=" + extension.toAbsolutePath().toString());
ChromeDriver driver = new ChromeDriver(options);
```

According to the release notes, this is also possible with Firefox in version 4.1.1, recently released. I tried to use the new feature in one of my tests as follows:

```
Path extension = Paths
        .get(ClassLoader.getSystemResource("web-extension").toURI());
FirefoxDriver driver = new FirefoxDriver();
driver.installExtension(extension, true);
```

But I get the following exception:

```
org.openqa.selenium.InvalidArgumentException: ...\target\test-classes\web-extension is an invalid path
Build info: version: '4.1.1', revision: 'e8fcc2cecf'
System info: host: 'DESKTOP-VPVQ23J', ip: '192.168.56.1', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_251'
Driver info: driver.version: FirefoxDriver
	at org.openqa.selenium.firefox.AddHasExtensions$1.installExtension(AddHasExtensions.java:82)
	at org.openqa.selenium.firefox.FirefoxDriver.installExtension(FirefoxDriver.java:276)
	at io.github.bonigarcia.webdriver.jupiter.ch05.caps.extensions.LoadExtensionFirefoxJupiterTest.setup(LoadExtensionFirefoxJupiterTest.java:46)
...

Caused by: java.nio.file.AccessDeniedException: ...\test-classes\web-extension
	at sun.nio.fs.WindowsException.translateToIOException(Unknown Source)
	at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
	at sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
	at sun.nio.fs.WindowsFileSystemProvider.newByteChannel(Unknown Source)
	at java.nio.file.Files.newByteChannel(Unknown Source)
	at java.nio.file.Files.newByteChannel(Unknown Source)
	at java.nio.file.Files.readAllBytes(Unknown Source)
	at org.openqa.selenium.firefox.AddHasExtensions$1.installExtension(AddHasExtensions.java:80)
	... 63 more
```

It seems that it is not possible to give the folder in which the extension (under development) is located. If I use the packaged extension (as a zip), I managed to make it work, but I would like to use the unpackaged folder.

Is it not possible to use an unpackaged extension in Firefox (like in Chrome with `--load-extension`)?

**Comment by titusfortner** [comment]
First, in Chrome it would be easier to use the provided [`extension` key](https://chromedriver.chromium.org/capabilities#h.p_ID_106) instead of an argument, which is accessed via [`addExtensions()`](https://github.com/SeleniumHQ/selenium/blob/trunk/java/src/org/openqa/selenium/chromium/ChromiumOptions.java#L136)

Right now, all the languages are just working with zipped files not directories.
4.1.1 works if you zipped the directory yourself and used that file as the argument.

I think an argument can be made that this should be Selenium's responsibility instead of the user's responsibility, but since this isn't something that blocks a user from accessing the temporary extension functionality, I think it belongs in a separate feature request.

**Comment by bonigarcia** [comment]
I think I found a problem in the current implementation of `installExtension()` for `FirefoxDriver` in Java. I use a method to zip the extension used in Firefox (see [example](https://github.com/bonigarcia/browserwatcher/blob/master/src/test/java/io/github/bonigarcia/browserwatcher/test/parent/BrowserParentTest.java)). When the extension contains folders with assets (e.g., icons, JavaScript libraries, CSS, etc., for example [this](https://github.com/bonigarcia/browserwatcher/tree/master/ext)), these assets are not loaded by the extension in the browser controlled with Selenium.

**Comment by bonigarcia** [comment]
I have just discovered that there was no problem in the implementation of `installExtension()` as I said in my previous post. The problem was in the method `WebDriverManager.zipFolder()` I used to compress the extension. Thanks to PR by @nickgaya, I managed to solve the issue, and now the assets within folders are correctly loaded in my tests. Thanks a lot, @nickgaya!

**Comment by titusfortner** [comment]
This was merged, will be available in 4.5
